[NativeAOT] Object synchronization method was called from an unsynchronized block of code.

Intermittent `Object synchronization method was called from an unsynchronized block of code.` exceptions.
<!-- Error message template  -->
### Known Issue Error Message

Fill the error message using [step by step known issues guidance](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssueJsonStepByStep.md).

<!-- Use ErrorMessage for String.Contains matches. Use ErrorPattern for regex matches (single line/no backtracking). Set BuildRetry to `true` to retry builds with this error. Set ExcludeConsoleLog to `true` to skip helix logs analysis. -->

```json
{
  "ErrorMessage": ["Object synchronization method was called from an unsynchronized block of code.", "ObjectHeader.Release"],
  "ErrorPattern": "",
  "BuildRetry": false,
  "ExcludeConsoleLog": false
}
```


<!-- Known issue validation start -->
 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=752522
**Result validation:** :warning: Validation could not be done without an Azure DevOps build URL on the issue. Please add it to the "**Build:** :mag_right:" line.
**Validation performed at:** 7/3/2024 2:40:39 AM UTC
<!-- Known issue validation end -->

<!--Known issue error report start -->

### Report
|Build|Definition|Step Name|Console log|Pull Request|
|---|---|---|---|---|
|[779167](https://dev.azure.com/dnceng-public/public/_build/results?buildId=779167)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/779167/logs/818)|dotnet/runtime#80154|
|[778002](https://dev.azure.com/dnceng-public/public/_build/results?buildId=778002)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/778002/logs/1095)|dotnet/runtime#106403|
|[777711](https://dev.azure.com/dnceng-public/public/_build/results?buildId=777711)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/777711/logs/297)|dotnet/runtime#106515|
|[776075](https://dev.azure.com/dnceng-public/public/_build/results?buildId=776075)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/776075/logs/302)|dotnet/runtime#106418|
|[774475](https://dev.azure.com/dnceng-public/public/_build/results?buildId=774475)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/774475/logs/258)|dotnet/runtime#106349|
|[774360](https://dev.azure.com/dnceng-public/public/_build/results?buildId=774360)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/774360/logs/772)|dotnet/runtime#105030|
|[773753](https://dev.azure.com/dnceng-public/public/_build/results?buildId=773753)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/773753/logs/970)|dotnet/runtime#106238|
|[771962](https://dev.azure.com/dnceng-public/public/_build/results?buildId=771962)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/771962/logs/846)|dotnet/runtime#104393|
|[770489](https://dev.azure.com/dnceng-public/public/_build/results?buildId=770489)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/770489/logs/977)|dotnet/runtime#106172|
|[770443](https://dev.azure.com/dnceng-public/public/_build/results?buildId=770443)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/770443/logs/790)|dotnet/runtime#106167|
|[769888](https://dev.azure.com/dnceng-public/public/_build/results?buildId=769888)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/769888/logs/414)|dotnet/runtime#106000|
|[767716](https://dev.azure.com/dnceng-public/public/_build/results?buildId=767716)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/767716/logs/819)|dotnet/runtime#106040|
|[767255](https://dev.azure.com/dnceng-public/public/_build/results?buildId=767255)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/767255/logs/406)|dotnet/runtime#106010|
|[767233](https://dev.azure.com/dnceng-public/public/_build/results?buildId=767233)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/767233/logs/959)|dotnet/runtime#106014|
|[766310](https://dev.azure.com/dnceng-public/public/_build/results?buildId=766310)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/766310/logs/341)|dotnet/runtime#105737|
|[765458](https://dev.azure.com/dnceng-public/public/_build/results?buildId=765458)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/765458/logs/360)|dotnet/runtime#105920|
|[765307](https://dev.azure.com/dnceng-public/public/_build/results?buildId=765307)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/765307/logs/1040)|dotnet/runtime#105866|
|[764609](https://dev.azure.com/dnceng-public/public/_build/results?buildId=764609)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/764609/logs/852)|dotnet/runtime#105595|
|[763406](https://dev.azure.com/dnceng-public/public/_build/results?buildId=763406)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/763406/logs/740)|dotnet/runtime#105842|
|[763381](https://dev.azure.com/dnceng-public/public/_build/results?buildId=763381)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/763381/logs/263)|dotnet/runtime#105840|
|[760921](https://dev.azure.com/dnceng-public/public/_build/results?buildId=760921)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/760921/logs/301)|dotnet/runtime#105695|
|[760268](https://dev.azure.com/dnceng-public/public/_build/results?buildId=760268)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/760268/logs/294)|dotnet/runtime#105695|
|[758871](https://dev.azure.com/dnceng-public/public/_build/results?buildId=758871)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/758871/logs/811)|dotnet/runtime#105394|
|[756714](https://dev.azure.com/dnceng-public/public/_build/results?buildId=756714)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/756714/logs/854)|dotnet/runtime#105308|
|[752522](https://dev.azure.com/dnceng-public/public/_build/results?buildId=752522)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/752522/logs/409)|dotnet/runtime#105346|
|[752390](https://dev.azure.com/dnceng-public/public/_build/results?buildId=752390)|dotnet/runtime|Build Tests|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/752390/logs/942)|dotnet/runtime#103752|

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[755415](https://dev.azure.com/dnceng-public/public/_build/results?buildId=755415)|dotnet/runtime|[nativeaot.SmokeTests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=755415&view=ms.vss-test-web.build-test-results-tab&runId=19151874&resultId=100028)|dotnet/runtime#105433|
|[752703](https://dev.azure.com/dnceng-public/public/_build/results?buildId=752703)|dotnet/runtime|[System.Text.Tests.EncodingTests.GetEncoding_FromProvider_ByEncodingName_WithDisallowedEncoding_Throws(encodingName: \"utf-7\", codePage: 65000)](https://dev.azure.com/dnceng-public/public/_build/results?buildId=752703&view=ms.vss-test-web.build-test-results-tab&runId=19070804&resultId=107874)|dotnet/runtime#105170|
|[751577](https://dev.azure.com/dnceng-public/public/_build/results?buildId=751577)|dotnet/runtime|[WasmTestOnChrome-ST-System.Reflection.Metadata.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=751577&view=ms.vss-test-web.build-test-results-tab&runId=19035898&resultId=147533)||
|[750924](https://dev.azure.com/dnceng-public/public/_build/results?buildId=750924)|dotnet/runtime|[WasmTestOnV8-ST-System.IO.FileSystem.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=750924&view=ms.vss-test-web.build-test-results-tab&runId=19015478&resultId=144761)||
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|0|4|30|
<!--Known issue error report end -->

Unfortunately, it seems this has not yet been fixed. Here's a fresh hit from today:

- Queue: `windows-x64 Debug NativeAOT`
- Job: https://dev.azure.com/dnceng-public/public/_build/results?buildId=741963&view=logs&j=15d34cf8-8614-5f8a-e364-3c281b6c41dd&t=2f724b17-e574-5ab4-a401-210ce585d93f&l=817
- Affected PR: https://github.com/dotnet/runtime/pull/104904
- Output:
```

       --- End of inner exception stack trace ---
       at System.Threading.Tasks.TaskReplicator.Run[TState](TaskReplicator.ReplicatableUserAction`1, ParallelOptions, Boolean) + 0x205
       at System.Threading.Tasks.Parallel.ForWorker[TLocal,TInt](TInt, TInt, ParallelOptions, Action`1, Action`2, Func`4, Func`1, Action`1) + 0x4ea
    --- End of stack trace from previous location ---
       at System.Threading.Tasks.Parallel.ThrowSingleCancellationExceptionOrOtherException(ICollection, CancellationToken, Exception) + 0x52
       at System.Threading.Tasks.Parallel.ForWorker[TLocal,TInt](TInt, TInt, ParallelOptions, Action`1, Action`2, Func`4, Func`1, Action`1) + 0x5eb
       at System.Threading.Tasks.Parallel.ForEachWorker[TSource,TLocal](IList`1, ParallelOptions, Action`1, Action`2, Action`3, Func`4, Func`5, Func`1, Action`1) + 0x1a8
       at System.Threading.Tasks.Parallel.ForEachWorker[TSource,TLocal](IEnumerable`1, ParallelOptions, Action`1, Action`2, Action`3, Func`4, Func`5, Func`1, Action`1) + 0x1a3
       at System.Threading.Tasks.Parallel.ForEach[TSource](IEnumerable`1, ParallelOptions, Action`1) + 0xaa
       at ILCompiler.RyuJitCompilation.CompileMultiThreaded(List`1) + 0x16e
       at ILCompiler.RyuJitCompilation.ComputeDependencyNodeDependencies(List`1) + 0x25b
       at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ComputeDependencies(List`1) + 0x61
       at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ComputeMarkedNodes() + 0xe1
       at ILCompiler.RyuJitCompilation.CompileInternal(String, ObjectDumper) + 0x55
       at ILCompiler.Compilation.ILCompiler.ICompilation.Compile(String, ObjectDumper) + 0x75
       at ILCompiler.Program.Run() + 0x41ba
       at ILCompiler.ILCompilerRootCommand.<>c__DisplayClass240_0.<.ctor>b__0(ParseResult) + 0x41c
       at System.CommandLine.Invocation.AnonymousSynchronousCliAction.Invoke(ParseResult) + 0x3f
       at System.CommandLine.Invocation.InvocationPipeline.Invoke(ParseResult) + 0x1b5
D:\a\_work\1\s\artifacts\bin\coreclr\windows.x64.Debug\build\Microsoft.NETCore.Native.targets(309,5): error MSB3073: The command ""D:\a\_work\1\s\artifacts\bin\coreclr\windows.x64.Debug\ilc-published\\ilc" @"D:\a\_work\1\s\artifacts\tests\coreclr/obj/windows.x64.Debug/Managed/nativeaot\SmokeTests\HardwareIntrinsics\X64Avx512\native\X64Avx512.ilc.rsp"" exited with code 1. [D:\a\_work\1\s\src\tests\nativeaot\SmokeTests\HardwareIntrinsics\X64Avx512.csproj] [D:\a\_work\1\s\src\tests\build.proj]

```

--- a/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/RuntimeImports.cs
+++ b/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/RuntimeImports.cs
@@ -650,6 +650,10 @@ namespace System.Runtime
         internal static extern int InterlockedCompareExchange(ref int location1, int value, int comparand);
 
         [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        [RuntimeImport(RuntimeLibrary, "RhpLockCmpXchg32")]
+        internal static extern unsafe int InterlockedCompareExchange(int* location1, int value, int comparand);
+
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
         [RuntimeImport(RuntimeLibrary, "RhpLockCmpXchg64")]
         internal static extern long InterlockedCompareExchange(ref long location1, long value, long comparand);
 

--- a/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Interlocked.cs
+++ b/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Interlocked.cs
@@ -23,6 +23,25 @@ namespace System.Threading
 #endif
         }
 
+        // This is used internally by NativeAOT runtime in cases where having a managed
+        // ref to the location is unsafe (Ex: it is the syncblock of a pinned object).
+        // The intrinsic expansion for this overload is exactly the same as for the `ref int`
+        // variant and will go on the same path since expansion is triggered by the name and
+        // return type of the method.
+        // The important part is avoiding `ref *location` in the unexpanded scenario, like
+        // in a case when compiling the Debug flavor of the app.
+        [Intrinsic]
+        internal static unsafe int CompareExchange(int* location1, int value, int comparand)
+        {
+#if TARGET_X86 || TARGET_AMD64 || TARGET_ARM64 || TARGET_RISCV64
+            return CompareExchange(location1, value, comparand); // Must expand intrinsic
+#else
+            // test readability of the location
+            _ = *location1;
+            return RuntimeImports.InterlockedCompareExchange(location1, value, comparand);
+#endif
+        }
+
         [Intrinsic]
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static long CompareExchange(ref long location1, long value, long comparand)

--- a/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Interlocked.cs
+++ b/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Interlocked.cs
@@ -1,6 +1,7 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
 using System.Runtime;
 using System.Runtime.CompilerServices;

--- a/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Interlocked.cs
+++ b/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Interlocked.cs
@@ -36,8 +36,7 @@ namespace System.Threading
 #if TARGET_X86 || TARGET_AMD64 || TARGET_ARM64 || TARGET_RISCV64
             return CompareExchange(location1, value, comparand); // Must expand intrinsic
 #else
-            // test readability of the location
-            _ = *location1;
+            Debug.Assert(location1 != null);
             return RuntimeImports.InterlockedCompareExchange(location1, value, comparand);
 #endif
         }

--- a/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/ObjectHeader.cs
+++ b/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/ObjectHeader.cs
@@ -152,7 +152,7 @@ namespace System.Threading
                 // there is nothing - try set hashcode inline
                 Debug.Assert((oldBits & BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX) == 0);
                 int newBits = BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX | BIT_SBLK_IS_HASHCODE | oldBits | newHash;
-                if (Interlocked.CompareExchange(ref *pHeader, newBits, oldBits) == oldBits)
+                if (Interlocked.CompareExchange(pHeader, newBits, oldBits) == oldBits)
                 {
                     return newHash;
                 }
@@ -247,7 +247,7 @@ namespace System.Threading
                 newBits = oldBits & ~(BIT_SBLK_IS_HASHCODE | MASK_HASHCODE_INDEX);
                 newBits |= syncIndex | BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX;
             }
-            while (Interlocked.CompareExchange(ref *pHeader, newBits, oldBits) != oldBits);
+            while (Interlocked.CompareExchange(pHeader, newBits, oldBits) != oldBits);
         }
 
         //
@@ -312,7 +312,7 @@ namespace System.Threading
                     // N.B. hashcode, thread ID and sync index are never 0, and hashcode is largest of all
                     if ((oldBits & MASK_HASHCODE_INDEX) == 0)
                     {
-                        if (Interlocked.CompareExchange(ref *pHeader, oldBits | currentThreadID, oldBits) == oldBits)
+                        if (Interlocked.CompareExchange(pHeader, oldBits | currentThreadID, oldBits) == oldBits)
                         {
                             return -1;
                         }
@@ -369,7 +369,7 @@ namespace System.Threading
                         if ((oldBits & MASK_HASHCODE_INDEX) == 0)
                         {
                             int newBits = oldBits | currentThreadID;
-                            if (Interlocked.CompareExchange(ref *pHeader, newBits, oldBits) == oldBits)
+                            if (Interlocked.CompareExchange(pHeader, newBits, oldBits) == oldBits)
                             {
                                 return -1;
                             }
@@ -398,7 +398,7 @@ namespace System.Threading
                             int newBits = oldBits + SBLK_LOCK_RECLEVEL_INC;
                             if ((newBits & SBLK_MASK_LOCK_RECLEVEL) != 0)
                             {
-                                if (Interlocked.CompareExchange(ref *pHeader, newBits, oldBits) == oldBits)
+                                if (Interlocked.CompareExchange(pHeader, newBits, oldBits) == oldBits)
                                 {
                                     return -1;
                                 }
@@ -458,7 +458,7 @@ namespace System.Threading
                             oldBits - SBLK_LOCK_RECLEVEL_INC :
                             oldBits & ~SBLK_MASK_LOCK_THREADID;
 
-                        if (Interlocked.CompareExchange(ref *pHeader, newBits, oldBits) == oldBits)
+                        if (Interlocked.CompareExchange(pHeader, newBits, oldBits) == oldBits)
                         {
                             return;
                         }

--- a/src/coreclr/nativeaot/Test.CoreLib/src/System/Runtime/RuntimeImports.cs
+++ b/src/coreclr/nativeaot/Test.CoreLib/src/System/Runtime/RuntimeImports.cs
@@ -95,6 +95,10 @@ namespace System.Runtime
         internal static extern int InterlockedCompareExchange(ref int location1, int value, int comparand);
 
         [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        [RuntimeImport(RuntimeLibrary, "RhpLockCmpXchg32")]
+        internal static extern unsafe int InterlockedCompareExchange(int* location1, int value, int comparand);
+
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
         [RuntimeImport(RuntimeLibrary, "RhpLockCmpXchg64")]
         internal static extern long InterlockedCompareExchange(ref long location1, long value, long comparand);
 

